### PR TITLE
Query subclient rework

### PR DESCRIPTION
**Pull request by timtay-microsoft**
Rework the v2 query subclient to lean more into getting results one at a time instead of one _page_ at a time. New usage looks like:

```csharp
using var serviceClient = new IotHubServiceClient("connString");
string queryText = "select * from devices";
QueryResponse<Twin> queryResults = serviceClient.QueryClient.CreateAsync(queryText);
while (await query.MoveNextAsync())
{
    // Get next result, may involve a service request to get the next page of results
    Twin twin = await query.Current;
    Console.WriteLine(twin);
}
```

Also add some e2e tests and remove the implementation files for v1 query.

Also added a reference to the query subclient from the jobs subclient so that the jobs subclient doesn't have to implement the query code locally.

### CODE REVIEW

File: e2e/test/iothub/service/QueryClientE2ETests.cs [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+using System.Threading.Tasks;
+using FluentAssertions;
+using Microsoft.Azure.Devices.E2ETests.helpers;
+using Microsoft.Azure.Devices.E2ETests.Helpers;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Azure.Devices.E2ETests.iothub.service
+{
+    /// <summary>
+    /// E2E test class for all registry operations including device/module CRUD.
```
Comment (timtay-microsoft) [inline e2e/test/iothub/service/QueryClientE2ETests.cs:17]:
```suggestion
    /// E2E test class for all Query client operations.
```

File: e2e/test/iothub/service/QueryClientE2ETests.cs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    /// </summary>
+    [TestClass]
+    [TestCategory("E2E")]
+    [TestCategory("IoTHub")]
+    public class QueryClientE2ETests : E2EMsTestBase
+    {
+        private readonly string _idPrefix = $"{nameof(QueryClientE2ETests)}_";
+
+        private readonly TimeSpan _timeout = TimeSpan.FromMinutes(1);
```
Comment (timtay-microsoft) [inline e2e/test/iothub/service/QueryClientE2ETests.cs:26]: More descriptive

File: e2e/test/helpers/RawQuerySerializationClass.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.Devices.E2ETests.helpers
```
Comment (drwill-ms) [inline e2e/test/helpers/RawQuerySerializationClass.cs:6]: helpers namespace casing

File: iothub/service/src/Query/QueriedPage.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+using System.Collections.Generic;
+using System.Net.Http;
+using Microsoft.Azure.Devices.Common.Extensions;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.Devices
+{
+    /// <summary>
+    /// Contains the result of a twin, scheduled job or raw query
```
Comment (drwill-ms) [inline iothub/service/src/Query/QueriedPage.cs:12]: punctuation

File: iothub/service/src/Query/QueryClient.cs [hunk 5]
```diff
@@ -61,18 +73,35 @@ internal QueryClient(string hostName, IotHubConnectionProperties credentialProvi
...
+
+                if (options?.PageSize != null)
+                {
+                    request.Headers.Add(PageSizeHeader, options.PageSize.ToString());
+                }
+
+                HttpResponseMessage response = await _httpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);
+                await HttpMessageHelper2.ValidateHttpResponseStatusAsync(HttpStatusCode.OK, response).ConfigureAwait(false);
+                string responsePayload = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                QueriedPage<T> page = new QueriedPage<T>(response, responsePayload);
```
Comment (drwill-ms) [inline iothub/service/src/Query/QueryClient.cs:103]: var

File: iothub/service/src/Query/QueryOptions.cs [hunk 6]
```diff
@@ -12,5 +12,10 @@ public class QueryOptions
         /// The token to use for continuing the query enumeration
```
Comment (drwill-ms) [inline iothub/service/src/Query/QueryOptions.cs:12]: Might need some nuance here, as in you won't need to use this if you just iterate items and don't care about pages.

File: iothub/service/src/Query/QueryResponse.cs [hunk 7]
```diff
@@ -1,45 +1,131 @@
...
+            if (ContinuationToken == null && queryOptions?.ContinuationToken == null)
+            {
+                // The query has no more pages of results to return and the last page has been
+                // completely exhausted.
+                return false;
+            }
+
+            // User's can pass in a continuation token themselves, but the default behavior
+            // is to use the continuation token saved by this class when it last retrieved a page.
+            var queryOptionsClone = new QueryOptions()
```
Comment (drwill-ms) [inline iothub/service/src/Query/QueryResponse.cs:96]: optional: ()
```suggestion
            var queryOptionsClone = new QueryOptions
```

File: iothub/service/src/Query/QueryResponse.cs [hunk 8]
```diff
@@ -1,45 +1,131 @@
...
         }
+
+        /// <summary>
+        /// Get the current query result. Can be called multiple times without advancing the query.
+        /// </summary>
+        /// <remarks>
+        /// Like with a more typical implementation of IEnumerator, this value is null until the first
+        /// <see cref="MoveNextAsync(QueryOptions, CancellationToken)"/> call is made.
+        /// </remarks>
+        public T Current { get; private set; }
```
Comment (drwill-ms) [inline iothub/service/src/Query/QueryResponse.cs:129]: move above methods, below ctors

File: iothub/service/src/Query/QueryResponse.cs [hunk 9]
```diff
@@ -1,45 +1,131 @@
...
+                CurrentPage = (IEnumerable<T>)response.CurrentPage;
+                Current = CurrentPage.GetEnumerator().Current;
+                ContinuationToken = response.ContinuationToken;
+            }
+
+            return true;
         }
+
+        /// <summary>
+        /// Get the current query result. Can be called multiple times without advancing the query.
```
Comment (drwill-ms) [inline iothub/service/src/Query/QueryResponse.cs:123]: Consider rewording. This looks like the item is the entire query result, not just the current enumerated item in the current page.

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**